cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -1256,6 +1256,66 @@ inline Status ConvertFixedSizeBinary(const ChunkedArray& data, PyObject** out_va
   return Status::OK();
 }
 
+inline Status ConvertStruct(const ChunkedArray& data, PyObject** out_values) {
+  PyAcquireGIL lock;
+  OwnedRef dict_item;
+  std::vector<OwnedRef> fields_data;
+  if (data.num_chunks() <= 0) {
+    return Status::OK();
+  }
+  // ChunkedArray has at least one chunk
+  auto arr = static_cast<StructArray*>(data.chunk(0).get());
+  // Use it to cache the struct type and number of fields for all chunks
+  auto num_fields = arr->fields().size();
+  auto array_type = arr->type();
+  for (int c = 0; c < data.num_chunks(); c++) {
+    auto arr = static_cast<StructArray*>(data.chunk(c).get());
+    fields_data.resize(num_fields);
+    // Convert the struct arrays first
+    for (size_t i = 0; i < num_fields; i++) {
+      PyObject* numpy_array;
+      RETURN_NOT_OK(ConvertArrayToPandas(arr->field(i), nullptr, &numpy_array));
+      fields_data[i].reset(numpy_array);
+    }
+
+    // Construct a dictionary for each row
+    const bool has_nulls = data.null_count() > 0;
+    for (int64_t i = 0; i < arr->length(); ++i) {
+      if (has_nulls && arr->IsNull(i)) {
+        Py_INCREF(Py_None);
+        *out_values = Py_None;
+      } else {
+        // Build the new dict object for the row
+        dict_item.reset(PyDict_New());
```

Comment:
I think using `OwnedRef` vs not using it still nets the same ref count.

* On the first iteration the refcount is 1
* You then increase it to 2 with `Py_INCREF(*out_values)` below
* On subsequent iterations the refcount of the object held by `dict_item` is 2, which decreases to 1 when you call `dict_item.reset(PyDict_New())`.
* The refcount just before the function exits is 2, but `OwnedRef`'s destructor decreases the refcount by 1 when it's called (when it goes out of scope).

Therefore, the net reference count of every `dict` is 1, which you can achieve without using `OwnedRef` or calling `Py_INCREF(*out_values)`. The code you had before was just fine :)